layer/filestore.go

```diff
@@ -98,6 +100,14 @@ func (fm *fileMetadataTransaction) SetCacheID(cacheID string) error {
 	return ioutil.WriteFile(filepath.Join(fm.root, "cache-id"), []byte(cacheID), 0644)
 }
 
+func (fm *fileMetadataTransaction) SetForeignSource(ref *distribution.Descriptor) error {
```

Comment:
Good idea. This means that you can't clear the foreign source with this, but this code didn't handle this case anyway.